java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory;
+
+/**
+ * A class for choosing the default allocation manager.
+ */
+public class DefaultAllocationManagerOption {
+
+  /**
+   * The environmental variable to set the default allocation manager type.
+   */
+  public static final String ALLOCATION_MANAGER_TYPE_ENV_NAME = "ARROW_ALLOCATION_MANAGER_TYPE";
+
+  /**
+   * The system property to set the default allocation manager type.
+   */
+  public static final String ALLOCATION_MANAGER_TYPE_PROPERTY_NAME = "arrow.allocation.manager.type";
+
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DefaultAllocationManagerOption.class);
+
+  /**
+   * The default allocation manager factory.
+   */
+  public static final AllocationManager.Factory DEFAULT_ALLOCATION_MANAGER_FACTORY =
+      getDefaultAllocationManagerFactory();
```

Comment:
what happens when logging during static initialization?

Reply:
If I understand your question correctly. You are concerned that when calling getDefaultAllocationManagerFactory, the logger may not have been initialized. If so, a NullPoinerException will be thrown. 

There is no need to worry about that. According to the Java specification (https://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.4.2), static members in a class are initialized in textual order, as though they were a single block. 

In our code, initialization of the logger appears before the DEFAULT_ALLOCATION_MANAGER_FACTORY, so it always gets initialized first.  

In fact, this code path has been tested by all our test cases, as by default, no environmental variable or system property are specified when running the tests.